client/src/main/java/io/pravega/client/stream/impl/Orderer.java

```diff
@@ -48,18 +53,39 @@
      */
     @VisibleForTesting
     public <T extends EventSegmentReader> T nextSegment(List<T> segments) {
+        return nextSegment(segments, 0L);
+    }
+
+    /**
+     * Similar to {@link Orderer#nextSegment(List)}, but adds the capability to block for data to arrive in the case
+     * that all the segments are idle. The maximum time to wait is defined via the timeout parameter (milliseconds).
+     *
+     * @param <T> The type of the SegmentInputStream that is being selected from.
+     * @param segments The logs to get the next reader for.
+     * @param timeout Timeout to wait for new data to arrive in the case that all segments are idle.
+     * @return A segment that this reader should read from next.
+     */
+    @SuppressWarnings("unchecked")
+    <T extends EventSegmentReader> T nextSegment(List<T> segments, long timeout) {
         if (segments.isEmpty()) {
             return null;
         }
+        List<CompletableFuture<T>> idleSegments = new ArrayList<>();
         for (int i = 0; i < segments.size(); i++) {
             T inputStream = segments.get(MathHelpers.abs(counter.incrementAndGet()) % segments.size());
             if (inputStream.isSegmentReady()) {
                 log.trace("Selecting segment: " + inputStream.getSegmentId());
                 return inputStream;
             } else {
-                inputStream.fillBuffer();
+                idleSegments.add(inputStream.fillBuffer().thenApply(b -> inputStream));
```

Comment:
This would effectively undo: https://github.com/pravega/pravega/pull/4082
That proved to be very significant optimization, and the change prevented the conditions needed to create #3900. 
I don't think we should do this.

Reply:
Perhaps a way around this, would be to add the stream to the idle segment list and then use the result of the read call to get the string of the segment and compare that to a map that you construct after the loop using the segmentid from the reader. 

That's not a great solution either though because it requires changing interfaces and makes the highest level code dependent on the wire protocol, but it also doesn't truly solve the problem in the event that the whole stream is idle. In that case we'll end up adding one completion stage per invocation.

Reply:
Maybe if we had something like a 'fillBufferAndSetListener' call which filled the buffer and set a listener to notify when it was completed instead of using a future. The trouble with this is I'm not sure how to make the set of the listener atomic with respect to the completion of the future. It's also not a great interface and it might cause problems with the threading model.